### ISSUE TITLE
Disable/improve thread affinity option

### ISSUE THREAD

**Issue by bnewton-intel** [intel/pcm#561]
After compiling from source on an Ubuntu 20.04.5 LTS system, I get the following error when executing ./pcm-sensor-server:

	=====  Processor information  =====
	Linux arch_perfmon flag  : yes
	Hybrid processor         : no
	IBRS and IBPB supported  : yes
	STIBP supported          : yes
	Spec arch caps supported : yes
	Max CPUID level          : 22
	CPU model number         : 85
	ERROR: pthread_setaffinity_np for core 1 failed with code 22
	PCM ERROR. Exception std::exception

My team believes this is due to another process on the system that already has core 1 pinned for its own needs. Is there a way to disable this feature, or can the binary determine a free core to use?

I edited the source code to disable thread affinity based on the comment in code that affinity is not implemented for OS X, recompiled, and the pcm-sensor-server appeared to work.

**Comment by ogbrugge** [comment]
22 is EINVAL on my systemy

Documentation says:
EINVAL (pthread_setaffinity_np()) The affinity bit mask mask
              contains no processors that are currently physically on
              the system and permitted to the thread according to any
              restrictions that may be imposed by the "cpuset" mechanism
              described in [cpuset(7)](https://man7.org/linux/man-pages/man7/cpuset.7.html).

       EINVAL (pthread_setaffinity_np()) cpuset specified a CPU that was
              outside the set supported by the kernel.  (The kernel
              configuration option CONFIG_NR_CPUS defines the range of
              the set supported by the kernel data type used to
              represent CPU sets.)

       EINVAL (pthread_getaffinity_np()) cpusetsize is smaller than the
              size of the affinity mask used by the kernel.

We'll need to look into this.

**Comment by bnewton-intel** [comment]
Thank you for your response.

I have gathered the following information:

```
prompt> lscpu
...
CPU(s):                          112
On-line CPU(s) list:             0-111
...

prompt> taskset -c 0 sleep 0
prompt> taskset -c 1 sleep 0
taskset: failed to set pid 3027471's affinity: Invalid argument
prompt> taskset -c 2 sleep 0
prompt> taskset -c 3 sleep 0
prompt> taskset -c 4 sleep 0
    
prompt> cat /sys/fs/cgroup/cpuset/cpuset.cpus
0-111
```

The process using cpu 1 has a cpuset file:

```
prompt> cat /sys/fs/cgroup/cpuset/myclient/cpuset.cpus
1,28,32,57,84,88

```

**Comment by rdementi** [comment]
in order to discover system topology (e.g. which socket core belongs to) and core type (E-core or P-core) to properly program core PMUs and a few other things pcm needs to execute cpuid instruction on each core. To run the instruction on each core pcm changes affinity of the thread. On OSX/Apple the thread affinity API is not available in user space therefore this logic is executed in pcm OSX driver. On other operating systems this is done in user-space. If one removes the affinity call the behavior of pcm is undefined (e.g. cores will be mapped to wrong sockets, wrong type of core on systems with hybrid CPUs, wrong events programmed, etc.). As a workaround for your issue would it be possible to assign PCM a cpuset that includes all cores? I see you were able to assign cores to "myclient".

**Comment by bnewton-intel** [comment]
I found the default /system cgroup cpuset has most of the cpus assigned but not the ones reserved for the "myclient" process:
0,2-27,29-31,33-56,58-83,85-87,89-111

The PCM binary has the error shown above when executed in this cgroup.

I created a "pcm" cpuset and assigned all cores to it. 
```
cgcreate -g cpuset:pcm
cgset -r cpuset.cpus=0-111 pcm
cgset -r cpuset.mems=0-3 pcm
```

I started a new instance of pcm-sensor-server assigned to that cgroup.
`cgexec -g cpuset:pcm --sticky ./pcm-sensor-server`

That allowed the binary to execute with no errors. 

But, some background process (possibly associated with "myclient", which I did not configure but is required on our system) comes around and updates the cpuset for the pcm cgroup, so it takes the 6 reserved cpus. As soon as that happens, the PCM binary crashes:
```
=====  Processor information  =====
...
Number of logical cores: 112
Number of online logical cores: 112
...
Starting plain HTTP server on http://localhost:9738/
# A few minutes pass, then:
ERROR: pthread_setaffinity_np for core 1 failed with code ERROR: pthread_setaffinity_np for core 57 failed with code 22
22
ERROR: pthread_setaffinity_np for core 28 failed with code 22
ERROR: pthread_setaffinity_np for core 84 failed with code 22
ERROR: pthread_setaffinity_np for core ERROR: pthread_setaffinity_np for core 88 failed with code 22
32 failed with code 22
terminate called after throwing an instance of 'std::exception'
  what():  std::exception
Aborted (core dumped)
```

**Comment by rdementi** [comment]
Thanks for the detailed reproducer. I can offer a workaround in pcm code where you can use your original cgroup setup where pcm will ignore the cores which it can not access/set affinity to (like they are offlined) excluding them from data collection.

**Comment by bnewton-intel** [comment]
I think that would be a great feature, we would definitely use that. Thank you!

**Comment by rdementi** [comment]
> I think that would be a great feature, we would definitely use that. Thank you!

it is now in the master branch

**Comment by bnewton-intel** [comment]
I have deployed the latest code with patch. I observed the following:

1. If I start the pcm-sensor-server process in a cpuset with a limited list of CPUs, and the cpuset does not change, the binary prints an error for each unavailable CPU and keeps running.
2. If I start the pcm-sensor-server process in a cpuset with all CPUs, and then a mechanism moves that process to the cpuset with limited CPUs, the binary in a short time will print an error for one of the CPUs that was previously available and then crash.

I have implemented the service using case 1 above and this is working well. Thank you.
